mdtraj/formats/pdbx/pdbxfile.py

```diff
@@ -0,0 +1,727 @@
+
+    def __init__(self, file):
+        """Load a PDBx/mmCIF file.
+
+        The atom positions and Topology can be retrieved by calling
+        getPositions() and getTopology().
+
+        Parameters
+        ----------
+        file : string
+            the name of the file to load.  Alternatively you can pass an open
+            file object.
+        """
+        top = Topology()
+        ## The Topology read from the PDBx/mmCIF file
+        self.topology = top
+        self._positions = []
+        self._unitcell_lengths = None
+        self._unitcell_angles = None
+
+        self._loadNameReplacementTables()
+
+        # Load the file.
+
+        inputFile = file
+        ownHandle = False
+        if isinstance(file, str):
+            inputFile = open_maybe_zipped(file, "r")
+            ownHandle = True
+        reader = PdbxReader(inputFile)
+        data = []
+        reader.read(data)
+        if ownHandle:
+            inputFile.close()
+        block = data[0]
+
+        # Build the topology.
+
+        atomData = block.getObj("atom_site")
+        atomNameCol = atomData.getAttributeIndex("auth_atom_id")
+        if atomNameCol == -1:
+            atomNameCol = atomData.getAttributeIndex("label_atom_id")
+        atomIdCol = atomData.getAttributeIndex("id")
+        resNameCol = atomData.getAttributeIndex("auth_comp_id")
+        if resNameCol == -1:
+            resNameCol = atomData.getAttributeIndex("label_comp_id")
+        resIdCol = atomData.getAttributeIndex("label_seq_id")
+        resNumCol = atomData.getAttributeIndex("auth_seq_id")
+        if resNumCol == -1:
+            resNumCol = resIdCol
+        resInsertionCol = atomData.getAttributeIndex("pdbx_PDB_ins_code")
+        chainIdCol = atomData.getAttributeIndex("auth_asym_id")
+        if chainIdCol == -1:
+            chainIdCol = atomData.getAttributeIndex("label_asym_id")
+            altChainIdCol = -1
+        else:
+            altChainIdCol = atomData.getAttributeIndex("label_asym_id")
+        if altChainIdCol != -1:
+            # Figure out which column is best to use for chain IDs.
+
+            idSet = set(row[chainIdCol] for row in atomData.getRowList())
+            altIdSet = set(row[altChainIdCol] for row in atomData.getRowList())
+            if len(altIdSet) > len(idSet):
+                chainIdCol, altChainIdCol = (altChainIdCol, chainIdCol)
+        elementCol = atomData.getAttributeIndex("type_symbol")
+        altIdCol = atomData.getAttributeIndex("label_alt_id")
+        modelCol = atomData.getAttributeIndex("pdbx_PDB_model_num")
+        xCol = atomData.getAttributeIndex("Cartn_x")
+        yCol = atomData.getAttributeIndex("Cartn_y")
+        zCol = atomData.getAttributeIndex("Cartn_z")
+        lastChainId = None
+        lastAltChainId = None
+        lastResId = None
+        lastInsertionCode = ""
+        atomTable = {}
+        atomsInResidue = set()
+        models = []
+
+        for row in atomData.getRowList():
```

Comment:
I'm glad this works, but some of this code is really wordy with many chained if-statements. Seems like it's going to be real difficult to maintain in the future.

At a minimum, I'm sure just spacing things with empty lines will help and adding in-line comments ABOVE the if statements to show what that next block is doing.